crates/storage/provider/src/providers/atomic.rs

```diff
@@ -0,0 +1,3507 @@
+use super::{DatabaseProviderRO, ProviderFactory, ProviderNodeTypes};
+use crate::{
+    providers::StaticFileProvider, AccountReader, BlockHashReader, BlockIdReader, BlockNumReader,
+    BlockReader, BlockReaderIdExt, BlockSource, ChainSpecProvider, ChangeSetReader, EvmEnvProvider,
+    HeaderProvider, ProviderError, PruneCheckpointReader, ReceiptProvider, ReceiptProviderIdExt,
+    StageCheckpointReader, StateReader, StaticFileProviderFactory, TransactionVariant,
+    TransactionsProvider, WithdrawalsProvider,
+};
+use alloy_eips::{BlockHashOrNumber, BlockId, BlockNumHash, BlockNumberOrTag, HashOrNumber};
+use alloy_primitives::{Address, BlockHash, BlockNumber, Sealable, TxHash, TxNumber, B256, U256};
+use parking_lot::RwLock;
+use reth_chain_state::{BlockState, CanonicalInMemoryState, MemoryOverlayStateProviderRef};
+use reth_chainspec::{ChainInfo, EthereumHardforks};
+use reth_db::models::BlockNumberAddress;
+use reth_db_api::models::{AccountBeforeTx, StoredBlockBodyIndices};
+use reth_evm::ConfigureEvmEnv;
+use reth_execution_types::{BundleStateInit, ExecutionOutcome, RevertsInit};
+use reth_primitives::{
+    Account, Block, BlockWithSenders, Header, Receipt, SealedBlock, SealedBlockWithSenders,
+    SealedHeader, StorageEntry, TransactionMeta, TransactionSigned, TransactionSignedNoHash,
+    Withdrawal, Withdrawals,
+};
+use reth_prune_types::{PruneCheckpoint, PruneSegment};
+use reth_stages_types::{StageCheckpoint, StageId};
+use reth_storage_api::{DatabaseProviderFactory, StateProvider, StorageChangeSetReader};
+use reth_storage_errors::provider::ProviderResult;
+use revm::{
+    db::states::PlainStorageRevert,
+    primitives::{BlockEnv, CfgEnvWithHandlerCfg},
+};
+use std::{
+    collections::{hash_map, HashMap},
+    ops::{Add, Bound, RangeBounds, RangeInclusive, Sub},
+    sync::Arc,
+};
+use tracing::trace;
+
+/// Type that interacts with a snapshot view of the blockchain (storage and in-memory) at time of
+/// instantiation.
```

Comment:
if we use the canonicalinmemory chain like this currently does, I don't think this is correct, because its content can be modified while this type exists?